rocAL/rocAL/source/readers/image/external_source_reader.cpp

```diff
@@ -0,0 +1,256 @@
+        bool ret = pop_file_data(image);
+        if (_end_of_sequence && !ret)
+        {
+            std::cerr<<"\n EOS || POP FAILED ";
+            return 0;
+        }
+        _file_data[_curr_file_idx] = image;
+        _current_file_size = std::get<1>(image);
+    }
+
+    return _current_file_size;
+}
+
+size_t ExternalSourceReader::read_data(unsigned char* buf, size_t read_size)
+{
+    if (_mode == FileMode::FILENAME) {
+        if(!_current_fPtr)
+            return 0;
+
+        // Requested read size bigger than the file size? just read as many bytes as the file size
+        read_size = std::min(static_cast<unsigned int>(read_size),_current_file_size);
+        size_t actual_read_size = fread(buf, sizeof(unsigned char), read_size, _current_fPtr);
+        return actual_read_size;
+    } else {
+        unsigned char *file_data_ptr = std::get<0>(_file_data[_curr_file_idx]);
+        size_t size = _current_file_size;
+        incremenet_read_ptr();
+        if (size > read_size)
+          THROW("Requested size doesn't match the actual size for file read")
+        memcpy(buf, static_cast<void *>(file_data_ptr), size);
+        return size;
+    }
+}
+
+void ExternalSourceReader::get_dims(int cur_idx, int& width, int& height, int& channels)
+{
+    if (cur_idx >= 0) {
+      width = std::get<2>(_file_data[cur_idx]);
+      height = std::get<3>(_file_data[cur_idx]);
+      channels = std::get<4>(_file_data[cur_idx]);
+    }
+}
+
+int ExternalSourceReader::close()
+{
+    return release();
+}
+
+ExternalSourceReader::~ExternalSourceReader()
+{
+    release();
+}
+
+int
+ExternalSourceReader::release()
+{
+    if (_mode != FileMode::FILENAME) {
+        if(!_current_fPtr)
+            return 0;
+        fclose(_current_fPtr);
+        _current_fPtr = nullptr;
+        _end_of_sequence = false;   // reset for looping
+    }
+    return 0;
+}
+
+
+void ExternalSourceReader::reset()
+{
+    if (_shuffle) std::random_shuffle(_file_data.begin(), _file_data.end());
+    _read_counter = 0;
+    _curr_file_idx = 0;
+    _end_of_sequence = false;   // reset for looping
+}
+
+size_t ExternalSourceReader::get_file_shard_id()
+{
+    if(_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    return _file_id  % _shard_count;
+}
+
+
+void ExternalSourceReader::push_file_name(const std::string& file_name)
+{
+    std::unique_lock<std::mutex> lock(_lock);
+    _file_names_que.push(file_name);
+    lock.unlock();
+    // notify waiting thread of new data
+    _wait_for_input.notify_all();
+}
+
+bool ExternalSourceReader::pop_file_name(std::string& file_name)
+{
+    std::unique_lock<std::mutex> lock(_lock);
+    if(_file_names_que.empty() && !_end_of_sequence)
+        _wait_for_input.wait(lock);
+    if (!_file_names_que.empty()) {
+      file_name = _file_names_que.front();
+      _file_names_que.pop();
+      return true;
+    }else
+      return false;
+
```

Comment:
Please remove blank line and adjust spacing after else at L212

Reply:
Done